### ISSUE TITLE
Using `experimentalOptimizeChunks` with static and dynamic imports of the same file

### ISSUE THREAD

**Issue by gingur** [rollup/rollup#3340]
- Rollup Version: 1.29.0
- Operating System (or Browser): Windows 10
- Node Version: 10.15.3

### How Do We Reproduce?

[Code REPL](https://rollupjs.org/repl/?version=1.29.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMGNvbW1vbkVudHJ5JTIwZnJvbSUyMCcuJTJGY29tbW9uRW50cnknJTNCJTVDbiU1Q25jb25zb2xlLmxvZygnZW50cnkxJTIwY29tbW9uRW50cnknJTJDJTIwY29tbW9uRW50cnkpJTNCJTVDbiU1Q25pbXBvcnQoJy4lMkZhc3luY0VudHJ5JykudGhlbihhc3luY0VudHJ5JTIwJTNEJTNFJTIwJTdCJTVDbiUyMCUyMGNvbnNvbGUubG9nKCdlbnRyeTElMjBhc3luY0VudHJ5JyUyQyUyMGFzeW5jRW50cnkpJTNCJTVDbiU3RCklM0IlNUNuJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIyZW50cnkyLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMGNvbW1vbkVudHJ5JTIwZnJvbSUyMCcuJTJGY29tbW9uRW50cnknJTNCJTVDbmltcG9ydCUyMColMjBhcyUyMGFzeW5jRW50cnklMjBmcm9tJTIwJy4lMkZhc3luY0VudHJ5JyUzQiU1Q24lNUNuY29uc29sZS5sb2coJ2VudHJ5MiUyMGNvbW1vbkVudHJ5JyUyQyUyMGNvbW1vbkVudHJ5KSUzQiU1Q25jb25zb2xlLmxvZygnZW50cnkyJTIwYXN5bmNFbnRyeSclMkMlMjBhc3luY0VudHJ5KSUzQiU1Q24lMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJhc3luYy5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJleHBvcnQlMjBkZWZhdWx0JTIwJ2FzeW5jRW50cnlDb250ZW50JyUzQiUyMiU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJub25FbnRyeUZpbGUuanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyZXhwb3J0JTIwZGVmYXVsdCUyMCdub25FbnRyeUZpbGVDb250ZW50JyUzQiUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJjanMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==)

Setting `experimentalOptimizeChunks` to true produces the following

### entry1.js
```js
'use strict';

var commonEntry = require('./asyncEntry-be86dbf9.js');

console.log('entry1 commonEntry', commonEntry.commonEntry);
new Promise(function (resolve) { resolve(require('./asyncEntry-3be3e543.js')); }).then(asyncEntry => {
  console.log('entry1 asyncEntry', asyncEntry);
});
```

### entry2.js
```js
'use strict';

var commonEntry = require('./asyncEntry-be86dbf9.js');

console.log('entry2 commonEntry', commonEntry.commonEntry);
console.log('entry2 asyncEntry', asyncEntry.asyncEntry$1);
```

### asyncEntry-3be3e543.js
```js
'use strict';

var asyncEntry = require('./asyncEntry-be86dbf9.js');



exports.default = asyncEntry.asyncEntry;
```

### asyncEntry-be86dbf9.js
```js
'use strict';

var commonEntry = 'commonEntryContent';

var asyncEntry = 'asyncEntryContent';

var asyncEntry$1 = /*#__PURE__*/Object.freeze({
	__proto__: null,
	'default': asyncEntry
});

exports.asyncEntry = asyncEntry;
exports.asyncEntry$1 = asyncEntry$1;
exports.commonEntry = commonEntry;
```

### Expected Behavior
The `entry1.js` file should async require `asyncEntry.js`, and `entry2.js` should statically require `asyncEntry.js`.

### Actual Behavior

In `entry2.js`, the static require statement for `asyncEntry.js` is completely missing. The identifier for `asyncEntry` is still used in the code but without the require statment it's resolving to a global that doesn't exist.

**Comment by lukastaegert** [comment]
Thank you for taking the time to create this issue. Unfortunately, `experimentalOptimizeChunks` is not only experimental but also largely untested and unmaintained (not to say unmaintainable). There have long been plans to remove it or replace it with a reimplementation. As the latter will take still quite some time, I will remove this feature for Rollup@2, see #3343 

It will most likely make a reappearance as a plugin in the future once the necessary changes and APIs have been created.

**Comment by frank-dspeed** [comment]
@lukastaegert don't do that let it stay its flaged experimental but it is usefull! i know that it comes from ember but still userfull don't remove else i need to fork i don't want that!

**Comment by lukastaegert** [comment]
But you are aware that it is broken as you can see above? And it is really hard to fix and maintain because it basically tries to incompletely replicate some logic without any connection to that logic. I would be interested to hear what the situation is that you are trying to improve with this feature. Also note that #3305 will land when this feature is removed, which can already noticeably decrease the number of chunks created by dynamic imports. There is another feature planned that will decrease the number of chunks created by static entry points by specifying that some entry points are always loaded before other entry points are loaded.

And of course there is `manualChunks`, which allows you to paint manually over the chunking today.

**Comment by frank-dspeed** [comment]
@lukastaegert i am well aware of this but let me show you a good example that it can get used 

this is a working workaround for the most bugs of it

```js
const rollup = require('../rollup');

rollup.rollup({
  experimentalCodeSplitting: true,
  optimizeChunks: true,
  input: ['main1.js', 'main2.js', 'main3.js']
}).then(build => {
  return build.generate({ format: 'amd' })
  .then(generated => {
    return build.generate({ format: 'amd' });
  });
}).then(generated => {
  console.log(generated);
})
.catch(e => {
  console.log(e);
})
``` 

this does exactly what you want to do in the new implamentation it moves it to the end of the generate phase.

**Comment by lukastaegert** [comment]
This is not what I mean. The chunking still happens in the `build` phase, `experimentalOptimizeChunks` just tries to merge chunks again in the generate phase. And since chunks are complicated objects, this merge is non-trivial and is now proven to be broken in edge cases.

I'm not sure what calling `generate` twice should accomplish except making everything slower (the optimization is only triggered during the first generate call).

Also, `experimentalCodeSplitting` has been deprecated for a year now.

**Comment by frank-dspeed** [comment]
hmm maybe your right and the functionality can be added via hooks anyway so the feature is not needed in the core project

**Comment by lukastaegert** [comment]
Closed in rollup@2.0.0
